### ISSUE TITLE
WLAN

### ISSUE THREAD

**Issue by assusdan** [supercairos/android_device_xiaomi_land#1]
Hello developer!
Any ideas about fixing wifi?

Thanks in advance!

**Comment by supercairos** [comment]
Did you managed to make it boot ?
For WLAN, just add prima module to your kernel.

**Comment by assusdan** [comment]
Yes I booted it.
I'm using stock kernel, so what I must do?

**Comment by supercairos** [comment]
Humm. Using the  Xiaomi Kernel ?

Add the  right prima open-source driver in the stagging folder of the
kernel.
If I haven't done it in my xiaomi kernel, have a look at my other kernel :)

I'm quite busy at the moment unfortunalty with work and won't have time to
dive into it before start of December.

Happy hacking :)

**Comment by assusdan** [comment]
Yes China Xiaomi kernel. Under right driver you mean .ko module? Can I use stock?

**Comment by yangyangnau** [comment]
The wlan issue probably is my fault.
My kernel and device tree (not commit yet) have wlan issue too.(It's ok to connect and ping route with static ip, but can't access Internet)
I will check stock kernel.

By the way:
@zwliew made another device tree with stock kernel. 
And @zwliew said "I have CM 13 almost daily driver material currently after a couple days of work."
https://github.com/zwliew/android_device_xiaomi_land

**Comment by supercairos** [comment]
Use this : https://source.codeaurora.org/quic/la/platform/vendor/qcom-opensource/wlan/prima/tree/?h=LA.UM.5.3_rb1.1 then place it in the stagging folder and edit the KConfig + Makefile files to make it compile. Add it as a module in the kernel defconfig and you should be good to go

**Comment by supercairos** [comment]
BTW @yangyangnau where did you find the Official Kernel from Xiaomi for Land ? I've failed to find it :(

**Comment by supercairos** [comment]
Btw, I didn't remember but it's added the WLAN driver to my own kernel : https://github.com/supercairos/android_kernel_xiaomi_msm8937/commit/9843f94c7e6afeb9078614699d46a352259b2a71

**Comment by assusdan** [comment]
@supercairos I made some changes in proprietary repo and in device tree (commits by vitek999 is my, I just want not set up my nickname in git), and replaced kernel in bootimage to stock and  /system/lib/modules to stock.

**Comment by assusdan** [comment]
You wanna say that if I'll compile your kernel with modules and pick .ko file WiFi will work? But I think it will cause module incompatibility error (uname -a not match)

**Comment by assusdan** [comment]
@yangyangnau I can't get how that device tree is working. :open_mouth:

**Comment by supercairos** [comment]
I would try to import my commit as a patch in your kernel and see if it's working. Again I had a lots of trouble with my kernel and a strange error "Too many open files" that caused my CM not to boot or to boot but being super slow. I was awaiting Xiaomi to release their Kernel source code. Trying to make our own kernel from QC Code Aurora source code is quite hard since some drivers are missings :(

**Comment by assusdan** [comment]
Custom kernel is so unstable, so I think we'll work on wifi in prebuilt kernel. I'll try to borrow wifi part form @zwliew 's tree. My testers (I have no Land) said our CM is more faster and smoother than his.

**Comment by yangyangnau** [comment]
@supercairos 
No official Kernel yet.

```
lynx -dump -width 120 https://wiki.codeaurora.org/xwiki/bin/QAEP/release | grep msm8937

mkdir -p android-source/qaep
cd android-source/qaep
repo init -u git://codeaurora.org/platform/manifest.git -b release -m LA.UM.5.3-01110-8x37.0.xml --repo-url=git://codeaurora.org/tools/repo.git
grep kernel/msm-3.18 .repo/manifest.xml
```

git://codeaurora.org/kernel/msm-3.18 branch LA.UM.5.3_rb1.1 985da67f18431ac30e2c2d46b68067b71ebf8688

My porting kernel based on LA.UM.5.3_rb1.1 and not LA.UM.5.1_rb1.4, I check it a few days ago.

**Comment by yangyangnau** [comment]
@assusdan @supercairos 
My porting kernel is stable enough with stock MIUI ROM. The wifi is OK.
(unless `adb connect ip`, and wait, `adb shell dmesg | tee /tmp/dmesg`will get "wifi net keeper" service crushing. But adb still work.)
(and some android app can't login using another app's user acc)

I will check @zwliew's device tree with my kernel.

**Comment by assusdan** [comment]
@yangyangnau maybe it's stable for your revision, but only 2 of 4 my testers successfully start it. Another one gets stuck at logo, and other has reboots between bootanimation and launcher start.

**Comment by yangyangnau** [comment]
@assusdan That's strange.
The Redmi 3S's stock MIUI ROM has enabled dm-virty default. So user need disable it.
If you are saying CM-13, my device tree isn't ready.

The wlan issue isn't kernel side issue from my tests.

**Comment by assusdan** [comment]
@yangyangnau We test on stock 6.11.03 global dev ROM.
Today I'll try wifi fixes.

**Comment by yangyangnau** [comment]
I tested @zwliew's device tree with my kernel.
It's unstable like my device tree.
The Wifi works, can connect at least.
Can't enter main menu yet, so I'm not sure if can access internet.

**Comment by ashishjohnson** [comment]
accually i had fixed my prima and compile working fine for santoni (redmi 4x) but while open  hotspot with mobile data on only working for one time , then wifi got stuck untill take a reboot

**Comment by ashishjohnson** [comment]
am using stock 3.18.31 and 3.18.31-perf
